tuna/miopen/celery_tuning/tuning.py

```diff
@@ -0,0 +1,390 @@
+    return True
+
+  res_set = ResultSet([])
+  start = time.time()
+
+  with DbSession() as session:
+    while True:
+      try:
+        job_list = []
+        #get all the jobs from mySQL
+        job_list = library.get_jobs(session, library.fetch_state,
+                                    library.set_state, library.args.session_id,
+                                    job_batch_size)
+
+        for i in range(0, len(job_list), job_batch_size):
+          batch_jobs = job_list[i:min(i + job_batch_size, len(job_list))]
+          if library.args.fin_steps:
+            entries = library.compose_work_objs_fin(session, batch_jobs,
+                                                    library.dbt)
+
+          entries = library.compose_work_objs_fin(session, job_list,
+                                                  library.dbt)
+          serialized_jobs = serialize_chunk(entries)
+
+          #build context for each celery task
+          for job, config in serialized_jobs:
+            context = {
+                'job': job,
+                'config': config,
+                'worker_type': worker_type,
+                'arch': library.dbt.session.arch,
+                'num_cu': library.dbt.session.num_cu,
+                'kwargs': kwargs,
+                'fdb_attr': fdb_attr
+            }
+
+            #calling celery task, enqueuing to celery queue
+            res_set.add(celery_enqueue.apply_async((context,), queue=q_name))
+
+        if not job_list:
+          if not res_set:
+            return False
+          LOGGER.info('All tasks added to queue')
+          break
+      except KeyboardInterrupt:
+        LOGGER.error('Keyboard interrupt caught, draining results queue')
+        session.rollback()
+        stop_active_workers()
+        results_gather(res_set, worker_type)
+        purge_queue([q_name], LOGGER)
+
+  results_gather(res_set, worker_type)
+  stop_active_workers()
+  end = time.time()
+  LOGGER.info("Took {:0>8}".format(str(timedelta(seconds=end - start))))  #pylint: disable=consider-using-f-string
+
+  return True
+
+
+def results_gather(res_set, worker_type):
+  """Start subproc to drain result queue and populate mysql DB"""
+  LOGGER.info('Started drain process')
+  #start draining result_queue in subprocess
+  drain_process = Process(target=drain_queue, args=[worker_type, DBT])
+  drain_process.start()
+
+  LOGGER.info('Gathering async results in callback function, blocking')
+  #waiting on all results to come in
+  _ = res_set.join(callback=result_callback)
+
+  #terminate result_queue drain process with special queue token (NONE,NONE)
+  LOGGER.info('Adding terminatig token to result_queue')
+  result_queue.put([None, None])
+
+  drain_process.join()
+  LOGGER.info('Done writing from result_queue to mySQL')
+
+
+def drain_queue(worker_type, dbt):
+  """Drain results queue"""
+  interrupt = False
+  LOGGER.info('Draining queue')
+  sleep_time = 0
+  wait_limit = 1800  #max wait time
```

Comment:
This (and lines 379-388) will terminate drain_queue after 30 minutes of accumulated deep sleep time?
Deep sleep being periods of over 2s where no work is completed (with an incremental use of 2s).
This seems generous, but I can also foresee the case where this could be running for a week and the accumulated delays would cause drain_queue to terminate before emptying the queue.
The bigger the workload, the higher the probability that all processors with find jobs that keep them busy for more than 2s.

Reply:
yes that was the intended behavior, to have a max wait time. As we discussed, I will find a different way to wait. I can up the wait time to something like 1 hour if that is more reasonable.

Reply:
If we replace this with a completion patience, that may work better.
We could reduce the wait_limit to 10 minutes, but refresh it every time a job is received by the result queue.
Then the behavior would be if no job completes in 10 minutes, exit the process.
I think this would reduce the probability of early termination proportionately.